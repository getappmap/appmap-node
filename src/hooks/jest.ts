import assert, { AssertionError } from "node:assert";
import { warn } from "node:console";
import { pathToFileURL } from "node:url";

import type { TransformTypes } from "@jest/types";
import { simple as walk } from "acorn-walk";
import type { ESTree } from "meriyah";

import * as gen from "../generate";
import globals from "../globals";
import genericTranform from "../transform";

export function shouldInstrument(url: URL): boolean {
  return url.pathname.endsWith("@jest/transform/build/ScriptTransformer.js");
}

export function transform(program: ESTree.Program): ESTree.Program {
  try {
    walk(program, { MethodDefinition });
  } catch (e) {
    assert(e instanceof AssertionError);
    warn(
      "Unknown Jest version. AppMap instrumentation cannot be applied.\n" +
        "Please report the problem and Jest version at https://github.com/getappmap/appmap-agent-js/issues\n",
      e,
    );
  }
  return program;
}

function MethodDefinition(method: ESTree.MethodDefinition) {
  const { key } = method;
  if (!isId(key, "transformSource")) return;

  const body = method.value.body;
  assert(body);

  walk(body, {
    ReturnStatement(ret: ESTree.ReturnStatement) {
      const arg = ret.argument;
      assert(arg);
      ret.argument = gen.call_(globals.transformJest, [
        gen.identifier("filename"),
        arg,
      ]);
    },
  });
}

function isId(node: ESTree.Node | null, name: string) {
  return node?.type === "Identifier" && node.name === name;
}

export function transformJest(
  filename: string,
  original: TransformTypes.TransformResult,
): TransformTypes.TransformResult {
  const xformed = genericTranform(original.code, pathToFileURL(filename));
  return { ...original, code: xformed };
}
