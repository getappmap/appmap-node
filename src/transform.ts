import assert from "node:assert";
import { writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { debuglog } from "node:util";
import { isNativeError } from "node:util/types";

import { generate } from "astring";
import { fromSource as getSourceMap, removeComments } from "convert-source-map";
import { parse, type ESTree } from "meriyah";

import applySourceMap from "./applySourceMap";
import * as instrument from "./hooks/instrument";
import * as jest from "./hooks/jest";
import * as mocha from "./hooks/mocha";
import { warn } from "./message";

const treeDebug = debuglog("appmap-tree");

export interface Hook {
  shouldInstrument(url: URL): boolean;
  transform(program: ESTree.Program): ESTree.Program;
}

const defaultHooks: Hook[] = [mocha, jest, instrument];

export function findHook(url: URL, hooks = defaultHooks) {
  return hooks.find((h) => h.shouldInstrument(url));
}

export default function transform(code: string, url: URL, hooks = defaultHooks): string {
  const hook = findHook(url, hooks);
  if (!hook) return code;

  try {
    const tree = parse(code, { source: url.toString(), next: true, loc: true, module: true });
    const sourcemap: unknown = getSourceMap(fixSourceMap(url, code))?.sourcemap;
    if (sourcemap) applySourceMap(tree, sourcemap);
    const xformed = hook.transform(tree);
    if (treeDebug.enabled) dumpTree(xformed, url);
    return generate(xformed);
  } catch (e) {
    assert(isNativeError(e));
    warn("Error transforming source at %s: %s", url, e.message);
    return code;
  }
}

function dumpTree(xformed: ESTree.Program, url: URL) {
  if (url.protocol !== "file:") return;
  const path = fileURLToPath(url) + ".appmap-tree.json";
  writeFileSync(path, JSON.stringify(xformed, null, 2));
  treeDebug("wrote transformed tree to %s", path);
}

// HACK: In node 18/16, when using --loader node-ts/esm
// sourcemap gets inserted twice to the typescript file. We remove the second one
// which reflects transpiled files map, instead of the original typescript file map.
function fixSourceMap(url: URL, code: string): string {
  const [major] = process.versions.node.split(".").map(Number);
  if (major >= 20) return code;

  if (![".ts", ".mts", ".tsx"].some((e) => url.pathname.toLowerCase().endsWith(e))) return code;

  let removed = removeComments(code);
  if (removed.indexOf("//# sourceMappingURL") > -1) {
    // Correct one remains, it needs to start in new line.
    removed = removed.replace("//# sourceMappingURL", "\n//# sourceMappingURL");
    return removed;
  }
  return code;
}
