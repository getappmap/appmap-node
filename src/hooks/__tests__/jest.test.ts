import console from "node:console";

import { parse } from "meriyah";
import * as jestHook from "../jest";
import transform from "../../transform";

describe(jestHook.shouldInstrument, () => {
  it("instruments ScriptTransformer", () => {
    expect(
      jestHook.shouldInstrument(
        new URL(
          "file:///test/node_modules/@jest/transform/build/ScriptTransformer.js",
        ),
      ),
    ).toBe(true);
    expect(
      jestHook.shouldInstrument(
        new URL(
          "file:///test/node_modules/@jest/transform/build/NotScriptTransformer.js",
        ),
      ),
    ).toBe(false);
  });
});

describe(jestHook.transform, () => {
  it("transforms the transformSource results", () => {
    const program = parse(`
      class ScriptTransformer {
        transformSource() {
          if (false) return 4;
          return 5;
        }

        otherMethod() {}
      }`);

    const xformed = jestHook.transform(program);

    expect(xformed).toEqual(
      parse(`
        class ScriptTransformer {
          transformSource() {
            if (false) return global.AppMap.transformJest(filename, 4);
            return global.AppMap.transformJest(filename, 5);
          }

          otherMethod() {}
        }`),
    );
  });

  it("generates a warning if code has unexpected structure", () => {
    const program = parse(`
      class ScriptTransformer {
        transformSource() { return; }
      }`);

    const warnSpy = (console.warn = jest.fn());

    expect(jestHook.transform(program)).toStrictEqual(program);
    expect(warnSpy.mock.lastCall).toMatchSnapshot();
  });
});

describe(jestHook.transformJest, () => {
  it("pushes jest transformed code through appmap hooks", () => {
    jest.mocked(transform).mockReturnValue("transformed test code");
    const result = jestHook.transformJest("/test/test.js", {
      code: "test code",
      originalCode: "original test code",
      sourceMapPath: null,
    });
    expect(result).toStrictEqual({
      code: "transformed test code",
      originalCode: "original test code",
      sourceMapPath: null,
    });
    expect(transform).toBeCalledWith(
      "test code",
      new URL("file:///test/test.js"),
    );
  });
});

jest.mock("../../transform");
