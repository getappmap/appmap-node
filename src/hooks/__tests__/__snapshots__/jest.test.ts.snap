// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform generates a warning if code has unexpected structure 1`] = `
[
  "Unknown Jest version. AppMap instrumentation cannot be applied.
Please report the problem and Jest version at https://github.com/getappmap/appmap-agent-js/issues
",
  [AssertionError: null == true],
]
`;
